src/Streamly/Internal/Data/Parse/Types.hs

```diff
@@ -0,0 +1,169 @@
+-- at a time, at any time the driver may @extract@ and examine the result of
+-- the fold. The fold may backtrack at any point, therefore, the driver holds
+-- the input in a backtracking buffer. The buffer grows or shrinks based on the
+-- return values of the fold step execution.
+--
+-- When a fold step is executed it generates a new state of the fold along with
+-- a command to the fold driver. The command tells the fold driver whether to
+-- keep the input for future backtracking or drop it. The constructors of
+-- 'Step' represent the commands to the driver.
+--
+-- /Internal/
+--
+data Step s =
+      Hold s     -- ^ Hold the current input in the buffer, the input is
+                 -- accepted tentatively, but the fold may later fail with the
+                 -- input remaining unused.
+    | Keep Int s -- ^ Keep @n@ most recent inputs, including the input of this
+                 -- step invocation, and discard the rest.  This indicates a
+                 -- commit of an intermediate result of the fold. A fold cannot
+                 -- fail beyond this commit point. If the fold terminates it
+                 -- will return at least the result up to this commit and the
+                 -- count of unused elements after that.
+    | Back Int s -- ^ Go back in the buffer by @n@ items, the count @n@
+                 -- includes the input element of this step as well. The fold
+                 -- driver backtracks and starts driving the fold with past
+                 -- inputs from the requested point in the buffer. Note that
+                 -- the driver cannot go back beyond the last commit point
+                 -- created by 'Keep'.
+    | Halt s     -- ^ Halt because the result of the fold will not change after
+                 -- this point.  Once a fold halts, driving it further may
+                 -- produce undefined behavior. An @extract@ after the 'Halt'
+                 -- would return a count of unused elements which includes the
+                 -- element on which 'Halt' occurred.
+
+-- Note: Instead of returning the unused count as part of extract, we can
+-- return it via Halt command. However, some folds may not ever halt, but they
+-- may still have unused input after the last commit point.
+
+-- | A parsing fold is represented as @Parse step initial extract@. Before we
+-- drive a fold we call the @initial@ action to retrieve the initial state of
+-- the fold. The driver invokes @step@ with the previous state and an input
+-- element. It results into a new state and a command to the driver represented
+-- by 'Step' type. At any point of time the driver can call @extract@ to
+-- inspect the result of the fold. It may result in an error or an output value
+-- along with the count (zero or more) of last unused elements.
+--
+data Parse m a b =
+    forall s. Parse (s -> a -> m (Step s))
+                    (m s)
+                    (s -> m (Either String (Int, b)))
```

Comment:
`Either String (Int, b)` does not seem right to me. How will you replay the input if you have an alternative fold? You need to include more formation if the parser fails.

Say, I want something like `takeExact 4 <|> takeExact 3`.

Reply:
I think I might be in the wrong here. Please don't reply to this yet.

Reply:
Please ignore this.